### PR TITLE
gh-320 Add OAuth backend infrastructure

### PR DESCRIPTION
**Pull request by ghillert**
resolves https://github.com/spring-cloud/spring-cloud-skipper/issues/320
resolves https://github.com/spring-cloud/spring-cloud-skipper/issues/321
resolves https://github.com/spring-cloud/spring-cloud-skipper/issues/322

### CODE REVIEW

**Comment by ilayaperumalg** [comment]
@ghillert Is there any reason why we can't use `spring-cloud-common-security-config` here?

**Comment by sabbyanandan** [comment]
@ghillert Please see Ilaya's comment ^^. The goal was to repurpose the security library; if it isn't used, let's circle back and discuss the reasons as to why.

**Comment by ghillert** [comment]
@ilayaperumalg Refactored to use as much as possible from Spring Cloud Common Security.

* Did NOT reuse `OAuthSecurityConfiguration` as the configuration is slightly different for Skipper
* Also `AuthorizationProperties` from `Spring Cloud Common Security` has the `enabled` property that is not used at all for Skipper

**Comment by ilayaperumalg** [comment]
> Did NOT reuse OAuthSecurityConfiguration as the configuration is slightly different for Skipper

@ghillert do you think we can move that difference as an extended class in the `common-config` and use it from the common configuration instead of having this class in Skipper?

Also, Could you elaborate on what are those differences, please?

**Comment by ghillert** [comment]
Some differences between Skipper and SCDF:

- One minor change is that AuthorizationProperties#enabled is not used any longer
- The browser detection is not needed any longer as we have now the `/api` and `/actuator` endpoints, meaning:
  - non- `/api` and `/actuator` endpoints will redirect to the OAuth Server, while `/api` and `/actuator` themselves will prompt for direct "basic authentication" or you provide an OAuth token. However, the REST endpoints will not redirect to the OAuth-server

**Comment by ilayaperumalg** [comment]
@ghillert Could you rebase the PR?

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/OAuthSecurityConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+		basicAuthenticationEntryPoint.setRealmName(securityProperties.getBasic().getRealm());
+		basicAuthenticationEntryPoint.afterPropertiesSet();
+		final Filter oauthFilter = oauthFilter();
+		BasicAuthenticationFilter basicAuthenticationFilter = new BasicAuthenticationFilter(
+				providerManager(), basicAuthenticationEntryPoint);
+		http.addFilterAfter(oauthFilter, basicAuthenticationFilter.getClass());
+		http.addFilterBefore(basicAuthenticationFilter, oauthFilter.getClass());
+		http.addFilterBefore(oAuth2AuthenticationProcessingFilter(), basicAuthenticationFilter.getClass());
+		this.authorizationProperties.getAuthenticatedPaths().add("/");
+		this.authorizationProperties.getAuthenticatedPaths().add("/api/**");
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/OAuthSecurityConfiguration.java]: These entries can go in `authenticatedPaths` settings.

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/OAuthSecurityConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+				http.authorizeRequests()
+						.antMatchers(this.authorizationProperties.getPermitAllPaths().toArray(new String[0]))
+						.permitAll()
+						.antMatchers(this.authorizationProperties.getAuthenticatedPaths().toArray(new String[0]))
+						.authenticated();
+
+		security = SecurityConfigUtils.configureSimpleSecurity(security, this.authorizationProperties);
+		security.anyRequest().denyAll();
+		this.securityStateBean.setAuthorizationEnabled(true);
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/OAuthSecurityConfiguration.java]: We can still use `authorizationProperties` and enable/disable based on the value.

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/OAuthSecurityConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	private ResourceServerProperties resourceServerProperties;
+
+	@Autowired
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	@Autowired
+	private AuthorizationProperties authorizationProperties;
+
+	@Override
+	protected void configure(HttpSecurity http) throws Exception {
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/OAuthSecurityConfiguration.java]: Based on the changes on this class, I believe we don't don't need to create the entire `OAuthSecurityConfiguration` class again here. Here are couple of possible options:

1) Extend the existing `OAuthSecurityConfiguration` and customize the `configure` method.
2) Modify `OAuthSecurityConfiguration` to inject `RequestMatcher`s like `BrowserDetectingContentNegotiationStrategy` programmatically and use that `OAuthSecurityConfiguration` here.

While we can go with option 1) for now, I can work on using Option 2) to make changes in the common security config.

In both ways, we can avoid re-creating this configuration class whole again.

**Comment by ghillert** [comment]
@ilayaperumalg Rebased and refactored.

Depends on https://github.com/spring-cloud/spring-cloud-common-security-config/issues/2

File: spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/SkipperOAuthSecurityConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+ * Setup Spring Security OAuth for the Rest Endpoints of Spring Cloud Data Flow.
+ *
+ * @author Gunnar Hillert
+ * @author Ilayaperumal Gopinathan
+ */
+@EnableOAuth2Client
+@EnableWebSecurity
+@Configuration
+@Conditional(OnSecurityEnabledAndOAuth2Enabled.class)
+public class SkipperOAuthSecurityConfiguration extends OAuthSecurityConfiguration {
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/config/security/SkipperOAuthSecurityConfiguration.java]: Now the configuration class looks super crisp :-)

File: spring-cloud-skipper-shell-commands/src/main/java/org/springframework/cloud/skipper/shell/config/HelpAwareShellApplicationRunner.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * {@link ApplicationRunner} that print a help-text to the console in case
+ * the help command-line option was provided and will also exit the application in that case.
+ *
+ * @author Gunnar Hillert
+ *
+ * @see DefaultShellApplicationRunner
+ * @see InitialConnectionApplicationRunner
+ */
+@Order(DefaultShellApplicationRunner.PRECEDENCE - 20)
+public class HelpAwareShellApplicationRunner implements ApplicationRunner {
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-shell-commands/src/main/java/org/springframework/cloud/skipper/shell/config/HelpAwareShellApplicationRunner.java]: This could possibly be on a separate PR as this is one other nice feature and can be tracked easily as a separate commit. Will try to separate out when merging, if not will be under the same one.

**Comment by ilayaperumalg** [comment]
@ghillert looks like there are compilation issues after the rebase:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project spring-cloud-skipper-shell-commands: Compilation failure: Compilation failure:
[ERROR] /Users/igopinathan/dev/git/spring-cloud/spring-cloud-skipper/spring-cloud-skipper-shell-commands/src/main/java/org/springframework/cloud/skipper/shell/config/ShellConfiguration.java:[63,129] cannot find symbol
[ERROR]   symbol:   class SkipperClientProperties
[ERROR]   location: class org.springframework.cloud.skipper.shell.config.ShellConfiguration
[ERROR] /Users/igopinathan/dev/git/spring-cloud/spring-cloud-skipper/spring-cloud-skipper-shell-commands/src/main/java/org/springframework/cloud/skipper/shell/command/support/InitializeConnectionApplicationRunner.java:[57,95] cannot find symbol
[ERROR]   symbol:   method isSkipSllValidation()
[ERROR]   location: variable skipperClientProperties of type org.springframework.cloud.skipper.client.SkipperClientProperties
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the ful
```

File: spring-cloud-skipper-docs/src/main/asciidoc/security.adoc [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+The rules regarding which REST endpoints require which roles, are specified in
+`application.yml` of the `spring-cloud-skipper-server-core` module.
+
+Nonetheless, you can override those, if desired. The configuration takes
+the form of a YAML *list* (as some rules may have precedence over others) and so
+you'll need to copy/paste the whole list and tailor it to your needs (as there is
+no way to merge lists). Always refer to your version of `application.yml`, as the snippet reproduced below may be out-dated. The default rules are as such:
+
+[source,yaml]
+----
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-docs/src/main/asciidoc/security.adoc]: I guess we wanted to show what's in the application.yml here. I can copy them over here.

File: spring-cloud-skipper-docs/src/main/asciidoc/security.adoc [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+      user-info-uri: http://127.0.0.1:9999/me
+----
+
+<1> Providing the Client Id in the OAuth Configuration Section will activate OAuth2 security
+
+You can verify that basic authentication is working properly using _curl_:
+
+[source,bash]
+----
+$ curl -u myusername:mypassword http://localhost:7577/
```
Comment (ilayaperumalg) [inline spring-cloud-skipper-docs/src/main/asciidoc/security.adoc]: This needs to be `http://localhost:7577/api`

**Comment by ilayaperumalg** [comment]
Addressing the latest comments along with #374 and merging. LGTM, also verified with OAuth server.

**Comment by ilayaperumalg** [comment]
Squashed and merged as 27d92bb75b095908aab1b7434ab438183ccb746c along with 7004b2c05d2fb6f2ada74019d341130e98808b94
